### ISSUE TITLE
Thoughts on a more concise system API

### ISSUE THREAD

**Issue by bfollington** [andoowhy/EgoCS#2]
Hey! I just stumbled across EgoCS after finding your initial reddit post about it. I've also looked at Entitas in the past and felt it was overkill for my needs, and I wanted to preserve the visual debugging Unity gives you in the inspector. So I really dig what you're doing here and how easy it is to get started.

I've just been playing around and I was wondering what you thought about a subclass of `EgoSystem` that circumvents the need to do:

```cs
public override void Update ()
{
    foreach( var bundle in bundles )
    {
        var transform = bundle.component1;
        var movement = bundle.component2;

         // ...move it by the velocity in its Movement Component
        transform.Translate( movement.velocity * Time.deltaTime );
    }
}
```

In favour of something more like:

```cs
public override void Update(Transform transform, Movement movement)
{
    transform.Translate( movement.velocity * Time.deltaTime );
}
```

For most of the cases I find myself in, I only want to work on one component bundle at a time. The code that facilitates this is (as you might expect) fairly trivial, but seems it would save a lot of time (for me at least).

Obviously, I can always just do this for myself, but I was wondering whether it might be worth including in the library? I'm unsure myself, as it isn't really core functionality, but I'd like your input :hand:.

**Comment by andoowhy** [comment]
I didn't go with that API at first to minimize EgoSystem method signature lengths. In my current game project, I often use empty Components as GameObject tags / flags, so I don't need to access them at all.

I'll think about it now that there's a demand for it.

**Comment by bfollington** [comment]
Ah yeah, that makes sense. Unfortunately (without some insanely redundant method signature generation) it's pretty hard to declare only the components you need as arguments. If we had access to new .NET features, you could actually get the argument type list and fulfil it automatically at runtime.

**Comment by benoneal** [comment]
I'm also in favour of removing this sort of boilerplate code. And really liking what you're doing here in general.

**Comment by bfollington** [comment]
:smile:

**Comment by andoowhy** [comment]
Now that I finally have `Start()`, `Update()`, and  `FixedUpdate()` done, I'm working on removing the `foreach( var bundle in bundles )` boilerplate when updating every System's GameObjects in Event Handlers. Although, it might actually be more code to write out :s

**Comment by bfollington** [comment]
I haven't really explored using events in Ego yet (haven't needed to) so I'm not really up to speed. When you say more code, do you mean for the end user or just in the system classes? I can see the method declarations getting very large for event handlers.

Out of interest, when you create the EgoSystem{X} classes are you using some sort of code generation? Just wondering, since they're very repetitive obviously.

**Comment by andoowhy** [comment]
For v0.3.0 and earlier, in an EgoSystem, if you wanted to iterate through all the GameObjects & its relevant Components, you'd have to write:

```C#
foreach( var bundle in bundles )
{
    var egoComponent = bundle.egoComponent;
    var myComponent1 = bundle.component1;
    var myComponent2 = bundle.component2;
    // Etc...

    ...
}
```

In v0.4.0, I added those special `Start( ... )`, `Update( ... )`, and `FixedUpdate( ... )` methods that removed the boilerplate, but in Event Handler methods, you would still need to write out the above `foreach( var bundle in bundles ) { ... }`. I didn't want two different ways to do the same thing.

So, in v0.5.0, I got rid of the special `Start( ... )`, `Update( ... )`, and `FixedUpdate( ... )` methods, *and* the `foreach` loop & boilerplate by creating the `ForEachGameObject( ... )` method, wherein you pass an appropriate lambda callback that gets invoked for each GameObject (the one time naming something wasn't hard \*phew\*). Now, you write the following:

```C#
ForEachGameObject( ( EgoComponent egoComponent, Transform transform, Rigidbody rigidbody ) =>
{
    ...
}
```
